### PR TITLE
[serve] Rolling updates for user_config changes

### LINKED ISSUE #15136: [serve] Support rolling updates for config changes
Now that we do rolling updates for code changes, we should also consider doing the same for config changes. These may be very slow (e.g., if it loads a new heavyweight model) which could cause downtime if all replicas do it simultaneously.

This should be straightforward to implement: just add another state and relevant transitions for `ReplicaState.UPDATING_CONFIG`.

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

Config updates may be slow and cause replicas to be unavailable while updating, so we should do rolling updates for these as well as code changes.

This PR uses the existing codepath for rolling updates but adds the hash of the user_config to the version used to check for required updates (this requires the user_config to be hashable, see `BackendVersion` implementation for details.

When the user-provided version is mismatched, we perform the existing rolling update. When only the user_config is mismatched, we all the reconfigure method and wait for it to return. This transitions the replica back to the `STARTING_OR_UPDATING` state.

Note that to maintain backwards compatibility this leaves the long poll client for backend configs in place, but once we remove the legacy `update_backend_config` codepath we can completely remove this from the backend_worker.

Note that https://github.com/ray-project/ray/pull/15820 is dependent on this change so that we can properly wait for config changes to go through (i.e., endpoints to be available).

## Related issue number

Closes https://github.com/ray-project/ray/issues/15136

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by simon-mo** [comment]
cc @caitengwei @liuyang-my This PR should help with common reloading models into running backend use case.

File: python/ray/serve/backend_state.py [hunk 1]
```diff
@@ -746,15 +834,12 @@ def _scale_backend_replicas(
             backend_info.backend_config.
             experimental_graceful_shutdown_timeout_s)
 
-        stopped = self._stop_wrong_version_replicas(
-            self._replicas[backend_tag], target_replicas, target_version,
-            graceful_shutdown_timeout_s)
-        if stopped > 0:
```
Comment (simon-mo) [inline python/ray/serve/backend_state.py:752]: intentional?

File: python/ray/serve/common.py [hunk 2]
```diff
@@ -24,16 +23,16 @@ class EndpointInfo:
...
-    actor_def: Optional[ActorClass]
-    version: Optional[str]
-    backend_config: BackendConfig
-    replica_config: ReplicaConfig
-
-    class Config:
-        # TODO(architkulkarni): Remove once ReplicaConfig is a pydantic
-        # model
-        arbitrary_types_allowed = True
+class BackendInfo:
```
Comment (simon-mo) [inline python/ray/serve/common.py:26]: dataclass?

**Review by simon-mo — APPROVED** [review]
LGTM overall, just few small comments.

File: python/ray/serve/backend_state.py (hunk 1 above)
Comment (edoakes) [reply python/ray/serve/backend_state.py:781]: yes this is now logged inside of the method

**Pull request merged**